Flesh out TensorPrimitives XML docs

stephentoub opened the pull request:
Contributes to https://github.com/dotnet/runtime/issues/92219

stephentoub committed:
Flesh out TensorPrimitives XML docs

stephentoub commented:
Note that in the doc comments, I noted that it's not safe to use an input span as an output span. I think that's a restriction we should remove, so that you can easily do in-place manipulations. We actually do already support this on some operations, but I left the comment on them for consistency. Now that the remainder processing support was merged, we can update the rest of the operations to make this safe, and then we can update those docs to clarify that overlaps between inputs and destination is safe iff they start at the same memory location.

stephentoub committed:
Merge branch 'main' into improvetensordocs

tannergooding commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:
```diff
@@ -6,72 +6,177 @@ namespace System.Numerics.Tensors
+        /// <paramref name="x"/> and <paramref name="destination"/> may not overlap; if they do, behavior is undefined.
+        /// </para>
+        /// <para>
+        /// This method may call into the underlying C runtime or employ instructions specific to the current architecture. Exact results may differ between different
```
Abs is one we can (and should) guarantee is the same everywhere (and which already should be consistent)

tannergooding commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:
```diff
@@ -6,72 +6,177 @@ namespace System.Numerics.Tensors
+        /// </para>
+        /// <para>
+        /// If a value is equal to <see cref="float.NegativeInfinity"/> or <see cref="float.PositiveInfinity"/>, the result stored into the corresponding destination location is set to <see cref="float.PositiveInfinity"/>.
+        /// If a value is equal to <see cref="float.NaN"/>, the result stored into the corresponding destination location is set to <see cref="float.NaN"/>.
```
Abs is a case where the spec requires NaN return the same bit pattern with the sign bit removed. It is one of the few cases we aren’t allowed to normalize

tannergooding replied:
Notably, there is never a reason to normalize for Abs either, since it’s more efficient to simply strip the sign bit and that’s what all platforms do already

tannergooding commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:
```diff
@@ -6,72 +6,177 @@ namespace System.Numerics.Tensors
+        /// If either of the element-wise input values is <see cref="float.NaN"/>, the resulting element-wise value is <see cref="float.NaN"/>.
+        /// </para>
+        /// <para>
+        /// This method may call into the underlying C runtime or employ instructions specific to the current architecture. Exact results may differ between different
```
Add, Subtract, Multiply, and Divide are likewise cases where all platforms should be trivially deterministic with no differences in result

we should never defer to the C runtime for them

tannergooding replied:
What might vary is the exact bit pattern for some NaN, as given NaN + NaN, it’s not guaranteed which NaN you get back

tannergooding commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:
```diff
@@ -6,72 +6,177 @@ namespace System.Numerics.Tensors
+        /// If any of the element-wise input values is <see cref="float.NaN"/>, the resulting element-wise value is <see cref="float.NaN"/>.
+        /// </para>
+        /// <para>
+        /// This method may call into the underlying C runtime or employ instructions specific to the current architecture. Exact results may differ between different
```
AddMultiply and MultiplyAdd are cases we should determine if we want to guarantee determinism or if we want to give the flexibility to use a fused operation internally, this would give a more accurate result across both operations and be within 1 ULP of the two separate operations

I’d lean towards determinism and that we should expose a separate MultiplyAddEstimate method which allows the optimization via explicit opt-in, as we’ve done elsewhere for math apis

stephentoub replied:
Should I interpret this as "we should revert https://github.com/dotnet/runtime/pull/92205"?

tannergooding replied:
CosineSimilarity is one of the more complex APIs where we can’t easily guarantee it today, so I think it’s fine there.

That’s different from the more primitive API, like MultiplyAdd, where it’s trivial to make the determinism guarantee and where we don’t really lose any perf by doing so.

tannergooding replied:
Long term, I’d like us to move towards cross platform determinism everywhere, but that’s a long goal and not something to worry about for this release where differences are already guaranteed to exist.

So I think we only need to be concerned with the places where we can trivially provide perf and determinism

stephentoub committed:
Address PR feedback

- Remove use of FusedMultiplyAdd from all but CosineSimilarity
- Remove comments about platform/OS-specific behavior from Add/AddMultiply/Subtract/Multiply/MultiplyAdd/Divide/Negate
- Loosen comments about NaN and which exact one is returned

stephentoub committed:
Merge remote-tracking branch 'upstream/main' into improvetensordocs

tannergooding commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:
```diff
@@ -6,72 +6,153 @@ namespace System.Numerics.Tensors
+        /// If a value is equal to <see cref="float.NaN"/>, the result stored into the corresponding destination location is also NaN.
+        /// </para>
+        /// <para>
+        /// The angles in x must be in radians. Multiply by <see cref="MathF.PI"/>/180 to convert degrees to radians.
```
In .NET 8, we have the new float.DegreesToRadians and inverse api

stephentoub committed:
Address PR feedback
